drivers/dma/dmamux_stm32.c

```diff
@@ -273,26 +349,50 @@ static const struct dma_driver_api dma_funcs = {
 #define DEV_DMA2 COND_CODE_1(DT_NODE_HAS_STATUS(DT_NODELABEL(dma2), okay), \
 			     DEVICE_DT_GET(DT_NODELABEL(dma2)), NULL)
 
+#define BDMA_1_BEGIN_DMAMUX_CHANNEL DT_PROP_OR(DT_NODELABEL(bdma1), dma_offset, 0)
+#define BDMA_1_END_DMAMUX_CHANNEL (BDMA_1_BEGIN_DMAMUX_CHANNEL + \
+				DT_PROP_OR(DT_NODELABEL(bdma1), dma_requests, 0))
+#define DEV_BDMA COND_CODE_1(DT_NODE_HAS_STATUS(DT_NODELABEL(bdma1), okay), \
+			     DEVICE_DT_GET(DT_NODELABEL(bdma1)), NULL)
+
 #define DEV_DMA_BINDING(mux_channel) \
 	((mux_channel < DMA_1_END_DMAMUX_CHANNEL) ? DEV_DMA1 : DEV_DMA2)
+
+#define DEV_BDMA_BINDING(mux_channel) \
+	(DEV_BDMA)
+
+
 #define DMA_CHANNEL(mux_channel) \
 	((mux_channel < DMA_1_END_DMAMUX_CHANNEL) ? \
 	 (mux_channel + 1) : (mux_channel - DMA_2_BEGIN_DMAMUX_CHANNEL + 1))
 
+#define BDMA_CHANNEL(mux_channel) \
+	((mux_channel < BDMA_1_END_DMAMUX_CHANNEL) ? \
+	 (mux_channel) : 0 /* not supported */)
+
 /*
- * No series implements more than 1 dmamux yet, dummy define added for easier
- * future extension.
+ * H7 series implements DMAMUX1 and DMAMUX2
+ * DMAMUX1 is used by DMA1 and DMA2
+ * DMAMUX2 is used by BDMA
+ *
+ * Note: Instance Number (or index) has no guarantee to which dmamux it refers
  */
-#define INIT_DMAMUX_0_CHANNEL(x, ...) \
-	{ .dev_dma = DEV_DMA_BINDING(x), .dma_id = DMA_CHANNEL(x), }
-#define INIT_DMAMUX_1_CHANNEL(x, ...) \
-	{ .dev_dma = 0, .dma_id = 0, }
 
-#define DMAMUX_CHANNELS_INIT_0(count) \
-	LISTIFY(count, INIT_DMAMUX_0_CHANNEL, (,))
-#define DMAMUX_CHANNELS_INIT_1(count) \
-	LISTIFY(count, INIT_DMAMUX_1_CHANNEL, (,))
+#define INIT_DMAMUX1_CHANNEL(x, ...) \
+	{ .dev_dma = DEV_DMA_BINDING(x), .dma_id = DMA_CHANNEL(x), }
+#define INIT_DMAMUX2_CHANNEL(x, ...) \
+	{ .dev_dma = DEV_BDMA_BINDING(x), .dma_id = BDMA_CHANNEL(x), }
+
+#if DT_SAME_NODE(DT_DRV_INST(0), DT_NODELABEL(dmamux1))
+#define INIT_INST0_CHANNEL(x, ...) INIT_DMAMUX1_CHANNEL(x, ...)
+#define INIT_INST1_CHANNEL(x, ...) INIT_DMAMUX2_CHANNEL(x, ...)
+#else
+#define INIT_INST0_CHANNEL(x, ...) INIT_DMAMUX2_CHANNEL(x, ...)
+#define INIT_INST1_CHANNEL(x, ...) INIT_DMAMUX1_CHANNEL(x, ...)
+#endif
```

Comment:
Non blocking note: I'm not quite fund of this specific part, but I see the reason.
One suggestion would be to review the macros defined above to take node id as argument, but that might be tricky.

Reply:
I agree this is not very clean, but I could not get it to work in a cleaner way. So I would be happy to mark this as op-of-scope for this PR if that's okay with you.